### PR TITLE
Add ValueTask factory members

### LINKED ISSUE #27960: API Proposal: Add to ValueTask: CompletedTask, FromResult and FromDefault
## Motivation

One of the original goals of `ValueTask` is handling functions that are "possibly async" in a zero-alloc fashion, for example:

```c#
public abstract ValueTask<Service> LocateServiceAsync(string service);

public override ValueTask<Service> LocateServiceAsync(string service)
{
   // Possibly async
   if (_cache.TryGetValue(service, out var result)) return new ValueTask<Service>(result);
   // Constructor is ambiguous for null, we have to repeatedly repeat ourselves.
   if (_nullCache.Contains(service)) return new ValueTask<Service>(default(Service));
   // Or:
   if (_nullCache.Contains(service)) return new ValueTask<Service>((Service)null);

   // Definitely async
   return Impl(service);
   async ValueTask<Service> Impl(string svc) { /* ... */ }
}
```

In Visual Studio this method is easy to type (as Intellisense supports `new` completion), but in VSCode and other editors the entirety of `ValueTask<Service>(service)` has to be typed out or copied from the method signature. The clarity and length of code also suffers, especially where any amount of nested generics are present (e.g. `Tuple` or `KeyValuePair`).

## Proposal

I propose adding the following to `ValueTask`, that follows from the static interface of `Task`:

```c#
public struct ValueTask
{
    // Improves consistency with regards to Task.
    // Prevents a Google search to determine if `return default` works (self-documenting).
    public static ValueTask CompletedTask { get; }

    public static ValueTask<T> FromResult<T>(T result);

    // Could be replaced with FromResult(default(T)), but that doesn't seem as elegant.
    public static ValueTask<T> FromDefault<T>();
}
```

## Result

This then allows generic parameter inference to work its magic, saving some keystrokes and greatly improving code clarity:

```c#
public override ValueTask<Service> LocateServiceAsync(string service)
{
   if (_cache.TryGetValue(service, out var result)) return ValueTask.FromResult(result);
   if (_nullCache.Contains(service)) return ValueTask.FromDefault<Service>();
   /* ... */
}
```

---

EDITED: 2/25/2020 by @stephentoub to paste proposal from below here, and update it with the implementation one-liners to show the semantics.
```C#
public struct ValueTask
{
    public static ValueTask CompletedTask => default;

    public static ValueTask<TResult> FromResult<TResult>(TResult result) =>
        new ValueTask<TResult>(result);

    public static ValueTask FromCanceled(CancellationToken cancellationToken) =>
        new ValueTask(Task.FromCanceled(cancellationToken));

    public static ValueTask<TResult> FromCanceled<TResult>(CancellationToken cancellationToken) =>
        new ValueTask<TResult>(Task.FromCanceled<TResult>(cancellationToken));

    public static ValueTask FromException(Exception exception) =>
        new ValueTask(Task.FromException(exception));

    public static ValueTask<TResult> FromException<TResult>(Exception exception) =>
        new ValueTask<TResult>(Task.FromException<TResult>(exception));
}
```

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes https://github.com/dotnet/runtime/issues/27960
cc: @tarekgh

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ValueTask.cs [hunk 1]
```diff
@@ -114,6 +111,40 @@ private ValueTask(object? obj, short token, bool continueOnCapturedContext)
             _continueOnCapturedContext = continueOnCapturedContext;
         }
 
+        /// <summary>Gets a task that has already completed successfully.</summary>
+        public static ValueTask CompletedTask => default;
```
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ValueTask.cs:115]: does default value mean completed?

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ValueTask.cs:115]: nevermind, I am seeing you already have a test case asserting default is completed task.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ValueTask.cs:115]: Yes, a `default(ValueTask)` is successfully completed.

**Review by tarekgh — APPROVED** [review]

**Comment by tarekgh** [comment]
@stephentoub 
```
System\IO\Compression\enc\BrotliStream.Compress.cs(79,20): error CS0173: (NETCORE_ENGINEERING_TELEMETRY=Build) Type of conditional expression cannot be determined because there is no implicit conversion between 'System.Threading.Tasks.ValueTask<int>' and 'System.Threading.Tasks.ValueTask'
```

**Comment by stephentoub** [comment]
Thanks.  Will fix.

File: src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ValueTask.cs (hunk 1 above)
Comment (bgrainger) [inline src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ValueTask.cs:115]: This member feels confusing from an API standpoint. Per https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.valuetask?view=net-5.0:

> The following operations should never be performed on a `ValueTask` instance:
> 
> * Awaiting the instance multiple times.

Although it's actually legal, it feels like using `ValueTask.CompletedTask` leads immediately to violating the above "should never" guideline. If I return this "instance" from my own method, won't my method's callers be awaiting the "same" instance multiple times and violating the rule? (Saying it's a by-value copy and they're actually different instances doesn't resolve the problem, as copying a `ValueTask` doesn't provide an exception to the current set of rules.)

If it's argued that this is a special case and can be treated differently, that just makes the ValueTask rules more difficult to comply with, as they can no longer be applied consistently.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ValueTask.cs:115]: Fair question, but that assumes it's returning the same instance each time you invoke the getter.  Think of it as a member you're invoking and it's giving you back an already completed task you can use once.  It's not special in that way.  I can invoke Stream.ReadAsync and use the returned ValueTask once.  I can invoke ValueTask.FromException and use the returned ValueTask once.  ValueTask.CompletedTask is the same.

Comment (bgrainger) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ValueTask.cs:115]: Thanks for the response; that's a fair interpretation.

I was going to argue that it is then inconsistent with `Task.CompletedTask`, which returns the same instance for efficiency, but then I read [the docs](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.completedtask?view=netcore-3.1) and saw that "Repeated attempts to retrieve this property value may not always return the same instance.", so TIL.

**Pull request merged**